src/react-redux.js

```diff
@@ -11,8 +11,18 @@ import type { BoundedDiff } from './generics';
 
 /* eslint-disable flowtype/generic-spacing */
 
+// Odd that $ReadOnly has any effect here -- the BoundedDiff just inside it
+// has one of its own.  But without it, Flow complains that C appears in
+// input/output position at ElementConfig<C>, incompatible with our marking
+// it as contravariant (so "input positions" only) with `-C`.  Either way,
+// that sure looks like a perfectly good input/contravariant position: it
+// goes through the top of two BoundedDiffs, then an ElementConfig, making a
+// path +, +, -, which multiplies to -.
+//
+// Meanwhile `-SP` is correct too: it goes through the top of one
+// BoundedDiff and the bottom of another, for a path +, -, multiplying to -.
```

Comment:
No, not category theory. I wouldn't so casually draw on that here, I hope :wink: That sentence is pretty shorthandy, though.

Here's the code for reference:
```js
export type OwnProps<-C, -SP, -D> = $ReadOnly<
  BoundedDiff<BoundedDiff<$Exact<ElementConfig<C>>, SP>, {| +dispatch: D |}>,
>;
```

Then what that thought ending in "multiplies to -" expands to is:
* That place where the type parameter `C` gets used in the definition of `OwnProps` is:
  * as the first argument to `ElementConfig`
  * … the result of which is used as the argument to `$Exact`
  * … the result of which is used as the first argument to `BoundedDiff`
  * … the result of which is used as the first argument to (another) `BoundedDiff`
  * … the result of which is used as the argument to `$ReadOnly`
  * … the result of which is the definition.
* The variance of each of those type-operators in their respective type parameters is:
  * contravariant, aka input, written `-`: `ElementConfig<-C>` (this isn't obvious, more on it below)
  * (… gloss over `$Exact`)
  * covariant, aka output, written `+`: `BoundedDiff<+U, -L>`, and this is being passed as that first parameter `U`
  * covariant, aka output, written `+`: `BoundedDiff<+U, -L>`, and this is again being passed as that first parameter `U`
  * (… gloss over `$ReadOnly`)
* So the sequence of variances is -, +, +, reading outward, from where `C` is used out to the definition.
  * Or +, +, -, reading inward from the definition to that use of `C`, which is the direction I used in that comment.
* And the mathematical signs -, +, + multiply together to make -. (For example, the numbers -1 * +1 * +1 multiply to -1.)
  * Which is just shorthand for: if you flip something backward an odd number of times, it ends up backward. Whereas if you flip it an even number of times, it ends up facing forward again.
  * And contravariance means that a subtyping relationship between parameters produces a subtyping in the opposite direction between results: S <: T means `F<T>` <: `F<S>`. Whereas covariance means it produces a subtyping in the same direction: `F<S>` <: `F<T>`.

Perhaps that's helpful? Also happy to discuss in the office tomorrow, where we'll have a whiteboard too.

---

Now, why do I say `ElementConfig` is contravariant?

… Well, maybe I shouldn't, actually -- I just tried an isolated test case, without all the other stuff around it that I was figuring was getting things confused here, and Flow still says it's an input/output position (aka invariant):
```js
type F<C>     = ElementConfig<C>; // ok
type Fout<+C> = ElementConfig<C>; // error
type Fin<-C>  = ElementConfig<C>; // error
```
Here's the same test case [on flow.org/try](https://flow.org/try/#0JYWwDg9gTgLgBAbzjAnmApnAogG3SdAOxgGEJCAzYAczgF84KoIQ4ByKdAQwGMY2A3ACghqDHABiAHhIA+OAoUBebHgLEylGjNkC4Aen1wIAa1FpMEiAFcYUgNRy4K3PiKlyVajr2G46KGYoc3EJYEIpAFonZ1U3DU9tOV8jAKChIA).

I think of it as roughly contravariant because a React component taking props of type `Props` is basically like a function `Props => React.Node`. In fact, if it's a function component, it's exactly that.  And for `P1 => React.Node` to be a subtype of `P2 => React.Node`, it's necessary and sufficient that P2 be a subtype of P1.